json/src/test/java/org/jdbi/v3/json/AbstractIssue2395Test.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.json;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+
+import org.jdbi.v3.core.Handle;
+import org.jdbi.v3.core.mapper.reflect.ConstructorMapper;
+import org.jdbi.v3.sqlobject.customizer.BindBean;
+import org.jdbi.v3.sqlobject.statement.SqlQuery;
+import org.jdbi.v3.sqlobject.statement.SqlUpdate;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public abstract class AbstractIssue2395Test {
+
+    Handle handle;
+    Dao dao;
+
+    protected abstract Handle getHandle();
+
+    @BeforeEach
+    void setUp() {
+        this.handle = getHandle();
+        // register a mapper for the wrapper class that contains a json field.
+        this.handle.registerRowMapper(Wrapper.class, ConstructorMapper.of(Wrapper.class));
+
+        this.dao = handle.attach(Dao.class);
+
+        this.handle.execute("create table test2395 (id integer not null primary key, name varchar(255), data json)");
+    }
+
+    @Test
+    @DisplayName("Test storing and retrieving a nested json field.")
+    void testIssue2395() {
+        final Wrapper data1 = new Wrapper(1, "alice", Collections.singletonList(new JsonData("test", "value", 4)));
+        final Wrapper data2 = new Wrapper(2, "bob", Collections.singletonList(new JsonData("toast", "words", 8)));
+
+        dao.write(data1);
+        dao.write(data2);
+
+        List<Wrapper> data = dao.readAll();
+        assertThat(data).containsExactly(data1, data2);
+
+        assertThat(dao.retrieve(1)).isEqualTo(data1);
+        assertThat(dao.retrieve(2)).isEqualTo(data2);
+    }
+
+    public interface Dao {
+
+        @SqlUpdate("insert into test2395 (id, name, data) values (:id, :name, :data)")
+        void write(@BindBean Wrapper bean);
+
+        @SqlQuery("select * from test2395")
+        List<Wrapper> readAll();
```

Comment:
If this issue is testing #2395 , I think this is slightly different. The OP asked about mapping `@Json List<T>` from a `[T, T]` json response, which is not the same as mapping `List<@Json T>` from a result set with many rows of `T`, which is what this test checks